### PR TITLE
Kshitij/dsg pickle experiment

### PR DESCRIPTION
**Pull request by kshitijsachan**
Can now pickle DSG objects. The MDP is not being pickled, so it needs to be added back to Option, SkillChaining, and DSG objects when loading pickle file. 

Certain MDPs will need to be pickled using the `dill` library because `dill` handles lambda functions properly.

### CODE REVIEW

File: simple_rl/agents/func_approx/ddpg/DDPGAgentClass.py [hunk 1]
```diff
@@ -126,12 +110,9 @@ def __setstate__(self, state_dictionary):
         self.critic.load_state_dict(state_dictionary["critic_state"])
         self.target_actor.load_state_dict(state_dictionary["target_actor_state"])
         self.target_critic.load_state_dict(state_dictionary["target_critic_state"])
-
         self.actor_optimizer.load_state_dict(state_dictionary["actor_optimizer"])
         self.critic_optimizer.load_state_dict(state_dictionary["critic_optimizer"])
 
-        self.replay_buffer = state_dictionary["replay_buffer"]
```
Comment (abagaria) [inline simple_rl/agents/func_approx/ddpg/DDPGAgentClass.py:133]: How are we setting and getting the replay buffer?

File: simple_rl/agents/func_approx/dqn/DQNAgentClass.py [hunk 2]
```diff
@@ -332,13 +332,10 @@ def get_batched_qvalues(self, states):
...
-                    inits = option.batched_is_init_true(states)
-                    # terms = np.zeros(inits.shape) if option.parent is None else option.parent.batched_is_init_true(states)
-                    terms = option.batched_is_term_true(states)
-                    action_values[(inits != 1) | (terms == 1), idx] = np.min(action_values) - 1.
-                except:
-                    pdb.set_trace()
+                inits = option.batched_is_init_true(states)
+                # terms = np.zeros(inits.shape) if option.parent is None else option.parent.batched_is_init_true(states)
+                terms = option.batched_is_term_true(states)
+                action_values[(inits != 1) | (terms == 1), idx] = np.min(action_values) - 1.
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dqn/DQNAgentClass.py]: Lets get rid of these changes that don't have to do with pickling

File: simple_rl/agents/func_approx/dqn/DQNAgentClass.py [hunk 3]
```diff
@@ -506,6 +503,26 @@ def __init__(self, action_size, buffer_size, batch_size, seed, device):
...
+    def __getstate__(self):
+        excluded_keys = ("experience", "memory")
+        state_dictionary = {x: self.__dict__[x] for x in self.__dict__ if x not in excluded_keys}
+
+        # need to convert to tuples because can't pickle NamedTuple
+        memory = [tuple(exp) for exp in self.memory]
+        state_dictionary["memory"] = memory
+        return state_dictionary
+
+    def __setstate__(self, state_dictionary):
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dqn/DQNAgentClass.py:519]: what about the policy and target networks? How are they being get and set?

File: simple_rl/agents/func_approx/dsc/DeepSkillGraphAgentClass.py [hunk 4]
```diff
@@ -136,7 +147,7 @@ def dsg_run_loop(self, episodes, num_steps, start_state=None, test_event=None, e
                 goal_state = self.mdp.get_position(self.mdp.sample_random_state())
                 self.dsc_agent.global_option_experience_replay(random_episodic_trajectory, goal_state=goal_state)
 
-            if episode > 0 and episode % 50 == 0 and args.plot_gc_value_functions:
+            if episode > 0 and episode % 50 == 0 and self.plot_gc_value_functions:
```
Comment (abagaria) [inline simple_rl/agents/func_approx/dsc/DeepSkillGraphAgentClass.py:150]: thanks for fixing this

**Review by abagaria — CHANGES_REQUESTED** [review]

File: simple_rl/agents/func_approx/ddpg/DDPGAgentClass.py (hunk 1 above)
Comment (kshitijsachan) [reply simple_rl/agents/func_approx/ddpg/DDPGAgentClass.py:133]: It's being set in the for loop in line 97 of the new code. We don't need to explicitly set it.

File: simple_rl/agents/func_approx/dqn/DQNAgentClass.py (hunk 3 above)
Comment (kshitijsachan) [reply simple_rl/agents/func_approx/dqn/DQNAgentClass.py:519]: This is the setter for `ReplayBuffer`, not the setter for `DQNAgentClass`. The policy networks are being set on line 481.

Also, let me know if there is a better way to reference another line. :)

**Review by abagaria — APPROVED** [review]
Looks great -- I think we can merge this now. Thanks, Kshitij!

**Pull request merged**